surefire-providers/surefire-junit-platform/pom.xml

```diff
@@ -136,6 +136,47 @@
                     </includes>
                 </configuration>
             </plugin>
+            <plugin>
+              <groupId>org.apache.maven.plugins</groupId>
+              <artifactId>maven-jar-plugin</artifactId>
+              <executions>
+                <execution>
+                  <id>api-jar</id>
+                  <phase>package</phase>
+                  <goals>
+                    <goal>jar</goal>
+                  </goals>
+                  <configuration>
+                    <classifier>api</classifier>
```

Comment:
Wasn't sure whether to classify this JAR with `api` or `spi`. 

In any case, this is simply a slim version of the JAR for service providers who otherwise do not depend on `surefire-junit-platform`

Reply:
This looks like a hack. Why you have introduced Apache SPI and you change the default behavior this way? JUnit5 SPI do not exist with same purpose? The plugin has configuration of class/methods. If somebody changes this in the SPI, then what you be valid? Configuration or SPI? **Pls describe very properly what you are aiming for in the JIRA in details.**

Reply:
There are other two SPI modules `surefire-extensions-spi` (for the forked JVM) and `surefire-extensions-api` (for the plugin process) but nothing is suited for this scenario of providers.
Can you explain how the customer would implement the interface `TestSelectorFactory` you have introduced? What types of sources can be filtered and how they can be configured?

Reply:
Absolutely. 

Here is my implementation for ArchUnit's [`FieldSelector`](https://github.com/TNG/ArchUnit/blob/2f3ff7dccbe3e3d8900604801fd545dd0c2e70d8/archunit-junit/junit5/engine-api/src/main/java/com/tngtech/archunit/junit/FieldSelector.java): 

```
package com.tngtech.archunit.junit.surefire;

import com.tngtech.archunit.junit.FieldSource;
import org.apache.maven.surefire.junitplatform.TestSelectorFactory;
import org.junit.platform.engine.TestSource;

public class FieldSelectorFactory implements TestSelectorFactory {
    @Override
    public boolean supports(TestSource testSource) {
        return testSource instanceof FieldSource;
    }

    @Override
    public String getContainerName(TestSource testSource) {
        return ((FieldSource) testSource).getClassName();
    }

    @Override
    public String getSelectorName(TestSource testSource) {
        return ((FieldSource) testSource).getFieldName();
    }
}
```

It solves the problem of adding support for applying the same filtering logic Surefire now applies to methods, to other test sources (`FieldSource` in this case). 

I've already verified providing this implementation of the new SPI in ArchUnit solves [the linked ArchUnit issue](https://github.com/TNG/ArchUnit/issues/641) for Surefire.

Reply:
Also, if I understand correctly, here: 

> The plugin has configuration of class/methods. If somebody changes this in the SPI, then what you be valid? Configuration or SPI?

You're asking what happens if someone overwrites the service to something else than `MethodSelectorFactory`. The answer is here: 

```
private Collection<TestSelectorFactory> loadSelectorFactories()
    {
        return StreamSupport.stream( ServiceLoader.load( TestSelectorFactory.class ).spliterator(), false )
            .collect( Collectors.toSet() );
    }
```

Basically, there's no overwriting. `MethodSelectorFactory` is always provided as the default implementation, and  extending libraries can only add new service implementations. 

Thus, for those users who do not extend this SPI in any way, the behavior of Surefire will be consistent with the behavior before this PR.